### ISSUE TITLE
[API Proposal]: DefaultInterpolatedStringHandler - expose buffer and reset

### ISSUE THREAD

**Issue by mgravell** [dotnet/runtime#110505]
### Background and motivation

Cross-reference https://github.com/dotnet/runtime/issues/110504

In APIs where it is desirable to efficiently pass in interpolated string values, for example as keys, there is not currently a mechanism to access a span-based implementation. `DefaultInterpolatedStringHandler` does *everything* required for such scenarios, but after append: only `ToStringAndClear` is exposed.

### API Proposal

[source](https://github.com/dotnet/runtime/blob/1d1bf92fcf43aa6981804dc53c5174445069c9e4/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DefaultInterpolatedStringHandler.cs)

``` diff
  public ref struct DefaultInterpolatedStringHandler
  {
-     internal void Clear() {...}
+     public void Clear() {...}

-     internal ReadOnlySpan<char> Text => ...
+     public ReadOnlySpan<char> Text => ...

+     // alternative suggestion for Text, to avoid locking implementation
+     public bool TryGetSpan(out ReadOnlySpan<char> value)
+     {
+        value = Text;
+        return true;
+     }
  }
```

### API Usage

```csharp
var key = (id: 42, region: "abc");
obj.SomeMethodTakingKey($"abc {key.id:000} in {key.region}");

    public void SomeMethodTakingKey(ref DefaultInterpolatedStringHandler key)
    {
       DoSomething(key.Text); // or key.TryGetSpan
       key.Clear();
    }
```

### Alternative Designs

- Manually implement your own interpolated string handler, duplicating all the logic from `DefaultInterpolatedStringHandler`
- Wrap a `DefaultInterpolatedStringHandler`, forwarding all the append methods, and use unsafe-accessor etc to invoke the `internal` APIs

### Risks

If the simple `Text` approach is used, it locks the implementation to a single buffer (not `ReadOnlySequence<char>` or similar); `TryGetSpan` leaves that open, with `ToStringAndClear` being a viable fallback.

Failure to call `Clear` could leak arrays (not a hard leak; the GC will still catch them, but: not ideal); that sounds like a "use the API correctly" problem; alternatively, perhaps the C# compiler itself could invoke the `Clear()` method (once it is `public`, and possibly by looking for an attribute to be sure) *after* the invoke, as a fallback, so that this is mitigated.

**Comment by colejohnson66** [comment]
Related: #110427

**Comment by mgravell** [comment]
lol, my team are troublemakers; good to see that both me and @BrennanConroy saw the problem re `Clear()`, and had similar thoughts

also cross-ref https://github.com/dotnet/runtime/issues/55390 which was previously closed; one of the objections there was the "exposing `Text` constrains the implementation" - I think the `TryGetSpan` approach helps make that more acceptable.

**Comment by mgravell** [comment]
Radical additional step, which would still require this API change; consider:

```
using System;

var key = (id: 42, region: "abc");
var cache = new SomeType();
cache.DoThing($"abc {key.id:000} in {key.region}");

class SomeType
{
    public void DoThing(string key) => throw new NotSupportedException();

    public void DoThing(ReadOnlySpan<char> key)
    {
        // this is just to show it is working; obvs we don't actually ToString usually
        Console.WriteLine($"Key: {key.ToString()}");
    }
}
```

At the moment this becomes:

``` c#
        // ...
        defaultInterpolatedStringHandler.AppendFormatted(valueTuple.Item2);
        someType.DoThing(defaultInterpolatedStringHandler.ToStringAndClear());
```

Wouldn't it be nice if the `ReadOnlySpan<char>` API took precedence, and the implementation became:

``` c#
        // ...
        defaultInterpolatedStringHandler.AppendFormatted(valueTuple.Item2);
        someType.DoThing(defaultInterpolatedStringHandler.Text);
        defaultInterpolatedStringHandler.Clear();
```

**Comment by Sergio0694** [comment]
Isn't this the same as https://github.com/dotnet/runtime/issues/55390? I tried making the same proposal years ago but it got rejected, there's the API noted at the end of the comments in that issue 🥲

**Comment by colejohnson66** [comment]
API review did agree to possibly adding a “try get text” method in the future. This proposal would allow things like this:

```csharp
try
{
    if (!dish.TryGetSpan(out ReadOnlySpan<char> span))
        span = dish.ToString();
    // use span
}
finally
{
    dish.Clear();
}
```

**Comment by mgravell** [comment]
@colejohnson66 yes, that's exactly what I meant by

> `TryGetSpan` leaves that open, with `ToStringAndClear` being a viable fallback.

**Comment by mgravell** [comment]
An alternative would be to expose the `Length` and a `CopyTo` method, allowing the caller to copy the data *out*, but that seems an additional extra copy

**Comment by mgravell** [comment]
@Sergio0694 yes exactly the same - I didn't see yours at the time. I guess this is just "more data points" in terms of "this API is really desirable, and here's concrete examples why", with the "if not" being: we manually copy `DefaultInterpolatedStringHandler`, or use unsafe-accessor.

**Comment by stephentoub** [comment]
> Radical additional step, which would still require this API change

This is interesting. @jaredpar, @333fred, timelines aside, is that feasible from a language perspective? Basically allowing an interpolated string to target type to a span, preferred over string, and using an exposed API on the default builder to get the span?

**Comment by Tornhoof** [comment]
If you modify the string one, can you do the same for the utf8 one? https://github.com/dotnet/runtime/blob/5ae9467d6ab7b31bf1cda35ca15ca5a2d21d3046/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8.cs#L351

The _pos field and _success field are internal, thus 
> Wrap a DefaultInterpolatedStringHandler, forwarding all the append methods, and use unsafe-accessor etc to invoke the internal APIs

is required there too if you don't want to rewrite your own.

**Comment by mgravell** [comment]
@stephentoub assuming it is restricted to "scoped" scenarios (so the span can't be snagged) the biggest problem I can see there is lifetime management - it would effectively need to work like `using`. Maybe it would be better to handle this in an explicit library way, for example:

```
using var content = SomeType.Create($"some {interpolated} string {here} with");
return obj.SomeMethodTakingSpan(content);
```

Where `Create` takes a `ref DefaultInterpolatedStringHandler`, and `SomeType` just wraps that and adds `Dispose() => value.Clear();` and an implicit static conversion operator to `ReadOnlySpan<char>`. Heck, maybe `SomeType` *is* `DefaultInterpolatedStringHandler`?

**Comment by mgravell** [comment]
Of course, at the language level, maybe the parallel there is an addition:

> If an interpolated string handler implements `IDisposable` or the `IDisposable` pattern, then interpolated string conversions will include a `try`/`finally` that involves the `Dispose()` method.

**Comment by mgravell** [comment]
actually, *assignment* to a string handler type already works; with the addition of `.Clear()` and `.Span`, this would work:

``` c#
using System;
using System.Runtime.CompilerServices;

int id = 42;
string name = "def";
DefaultInterpolatedStringHandler value = $"abc {id} def {name}";
DoSomething(value.Text); // this .Text does not currently work
value.Clear(); // this .Clear() does not currently work

static void DoSomething(ReadOnlySpan<char> value){}
```

The only question there is: should the public-facing `Clear()` actually be `Dispose()`, which would allow:

``` c#
int id = 42;
string name = "def";
DefaultInterpolatedStringHandler value = $"abc {id} def {name}";
DoSomething(value.Text); // this .Text does not currently work
```

Demonstration that this does the right thing, including not creating a shadow copy of `value` for the `using` [via sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0ASgK4B2ALgJYC2MlAwhKwA7MAbGFADKwgG7MwMAM4BuQoWZNSzACakAvKRRYFRHGXoBDdltIAiNTABmF/SVIBZAJ4jGUZQHMAEsfpqQlCk4sYCtDDmACQWxsBgpADe6gC+pNY2SSbsKfaEACIQIjwwjAAW3gAUoeEcACowCIwAlPqEFEjkKKSFxezlVdQwxmoA8vQCbrz+ADxgZcZQAHwhYRHNSaQpigQ7ANoAkkzCvBACxowwau6e9L7+gcIAuu0AzKSwmTIetGCMzm4PN4/AEgpVlP8BMxLlAwgAZGB3coAGlUKhs0FYF0uam4DBahEShFIJPI73ytmMtAEjCOMNO5xxN2BDyCpAWoOE5noMAA7pUoTD4YivCjSBioFjGDi8UxWsTSRh3hhuvlmDJTjIYJUNpoVpwhItKp92azhPKCKSyaQhiNxpMRNN6HMFstSA0mloVgBxUoexjG2ykACq9BkxhsHAAgjIhjZwfRTZyoM0LVaAPTp0gAPVzIbDEciZWEkXVpH8aJq6nS/i8wggtBkprAAGty1AGwFSJwRAAOV4oADs6Z7vZwg+I6dKYB2Vr2ofDkajYGkMhk0EqC8Ly9X66gAGllGpKE5SmUIGpUQA5UyRbQWA3DKAWZovS2k3ieUKXcg4TqNGFExVbtDSgQNMgpGwqRpOkTjOC4rmZO4QUeYIanWNp3xJecCyXFdZD3TdcJgHcCOgQ8AhPM8L2vW9zAsOtGAAfX9F83ytDpSAA4RE1tMYJimWZ5kWH0/QA8CekpalaWOKAGQQ64gWQs00LWGALQVEklS6Ugo14XhETUOFoWEMJKgoMh0PUr0k1Qyg9IMgJjKFARqjUjSsOtYCHMMgAxTFsSuczDAAflWWpdRWDk7J8gJ/MlQK1DciL9A45VulitR4qlHFKj4+1BOdYS3SsyLbKCez9L8gLpSC0rUsVdLdKquKapxGY6iWSo6nC9YbOiirMuyxKOq6+rNK8jKWqytqrlG7repgVEIXLKEvHodg5X6lTKscmaEtqtR5qs1EwmYdbNpaBqtKaobZqOzqFpO39iDCiUpTKgbhF26qDvax7nvei400anS7r+ubHp656VrOi7EUYVELLemrPp28Gcshsa1NOtaNoR1EgauibMy4xgZwIFIgA==)

**Comment by mgravell** [comment]
> some folks even argue for not doing so

I've started using that approach almost exclusively; the reason being: a lot of code is async, and if an exception happens in async: *we don't know* whether the buffer is still being touched by a competing code path, and we're seeing a CTS that has been cancelled by a CT.

**Comment by stephentoub** [comment]
> and if an exception happens in async: we don't know whether the buffer is still being touched by a competing code path

You don't know that for sync either, e.g. if code does:
```C#
public int Compute(byte[] buffer)
{
    Task t1 = Task.Run(() => Compute(buffer.AsSpan(0, buffer.Length / 2));
    Task t2 = Task.Run(() => Compute(buffer.AsSpan(buffer.Length / 2));
    return t1.Result + t2.Result;
}
```
The point being, it's not really about the signature being sync vs async but rather about whether the implementation uses fork/join patterns. If an async method doesn't fork (the vast vast majority don't), or if it guarantees that all forked work will have quiesced (e.g. Task.WhenAll), then there's no difference in this regard.

But we digress...

**Comment by Sergio0694** [comment]
> "Demonstration, although notes that it creates a shadow copy of value for the `using`"

This makes me think back of the `moveonly` proposal we (Jared, Jan, Stephen, Tanner, and I) talked about a while back. Didn't go anywhere because there was not too much interest for it at the time. I wonder if perhaps this could be a good occasion to try to get more traction for it (also considering the recent push for security, and copies being an easy way to introduce security problems when array pooling is involved). More of a side point I suppose, but I'd be happy to ping that thread again and work with folks to write an actual spec to bring to language review, if there's interest for that too now 🙂

**Comment by mgravell** [comment]
Here's a fully worked example of how two proposed tweaks combine to allow complex cache scenarios to use alloc-free code paths, including retroactively when a string API already exists: https://gist.github.com/mgravell/750bed134f36f417f97c217297552a88

**Comment by TrayanZapryanov** [comment]
@mgravell Is it an alternative also to build a custom InterpolatedStringHandler with StringBuilderCache as a storage?
Then There is no need of exposing new API and you will receive almost same performance characteristics, but it will be much safer than somebody forget to call Clear() and did not return baking array to the pool? I didn't measure how much slower is the usage of StringBuilderCache against something like ValueStringBuilder, but it is safer in my eyes.

@stephentoub FYI

**Comment by TrayanZapryanov** [comment]
@mgravell The idea is to get span from StringBuilder - I think [this can be used](https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.getchunks?view=net-9.0). And then it will be zero alloc too.

**Comment by mgravell** [comment]
My gut says "that won't be as efficient", but Eric tells me I should race my horses, so:

``` txt
| Method                             | Mean     | Error    | StdDev   | Gen0   | Allocated |
|----------------------------------- |---------:|---------:|---------:|-------:|----------:|
| Tuple                              | 11.18 ns | 0.047 ns | 0.041 ns |      - |         - |
| String                             | 35.17 ns | 0.176 ns | 0.156 ns | 0.0033 |      56 B |
| DISH_Span                          | 34.90 ns | 0.203 ns | 0.190 ns |      - |         - |
| StackAlloc_DISH_Span               | 32.61 ns | 0.155 ns | 0.145 ns |      - |         - |
| StackAllocSkipLocalsInit_DISH_Span | 32.31 ns | 0.101 ns | 0.090 ns |      - |         - |
| SBC_Span                           | 34.60 ns | 0.251 ns | 0.235 ns |      - |         - |
```

where

- `Tuple` is using `TKey` === `(string region,int id)`
- `String` is using `TKey` === `string` with `$"/foos/{region}/{id}"
- `DISH_Span` is using the same data as `String` via alt-lookup with the API proposed this ticket
- `StackAlloc_DISH_Span` is like `DISH_Span` but using `stackalloc` to provide `initialBuffer` instead of leaning on the array-pool
- `StackAllocSkipLocalsInit_DISH_Span` is like `StackAlloc_DISH_Span` but plus `[SkipLocalsInit]` for the `stackalloc`
- `SBC_Span` is using the suggestion from @TrayanZapryanov 

Conclusion: the zero-alloc string-like approaches are all within rounding-error distance of each-other; the alloc-version is *also* in the same field, but: we'd love to avoid that unnecessary alloc. It feels like this is a natural and useful API for `DefaultInterpolatedStringHandler`, where-as I've had to hack it in for `StringBuilderCache`. Additionally:

- `StringBuilderCache` isn't a public API; it would be per-consumer
- you'd need to add and maintain a *lot* of `AppendFormatted` overloads on your custom type to retain feature parity with `DefaultInterpolatedStringHandler`

[code for the above benchmark](https://gist.github.com/mgravell/0e49023b2ed41a255a2a94c3f6b4f320)

**Comment by TrayanZapryanov** [comment]
@mgravell Thank you very much for spending time to test my proposal.
I agree that it is not a perfect, just eliminates the need of changing existing interface.
According to the problems that you point in using it :
**StringBuilderCache isn't a public API; it would be per-consumer**
It is quite simple class and you should be able to mimic it in the handler or just clone and tune it for your use case

**you'd need to add and maintain a lot of AppendFormatted overloads on your custom type to retain feature parity with DefaultInterpolatedStringHandler**
This is not exactly true as StringBuilder already had a lot of overrides

![Image](https://github.com/user-attachments/assets/a0fd7d47-317f-419f-8604-bd2b33b36b80)
I guess compiler will find the correct one if you pass formattable value.

Again - I am not against your approach, just an idea that popup in my mind when I saw pending PR.

**Comment by stephentoub** [comment]
> I guess compiler will find the correct one if you pass formattable value.

No. The compiler requires very specific constructors and named methods.

> It is quite simple class

It's not, though. When do you remove a builder from the cache, and if you don't remove it, how do you avoid corruption if the same builder is used by two different consumers at the same time? How do you avoid allocation for larger strings without the cache growing too large, and if you allow it to grow large, how do you avoid that being a process-wide leak? Etc

**Comment by mgravell** [comment]
> This is not exactly true as StringBuilder already had a lot of overrides

My point is that you'd still need to offer them up to the compiler, though, via `AppendFormatted` (and `AppendLiteral`) methods on your wrapper that is marked `[InterpolatedStringHandler]`. To do that efficiently, when each item optionally taking `alignment` and `format` in addition to `value`: quickly becomes non-trivial.

**Comment by TrayanZapryanov** [comment]
Thank you both for the information.
As always very helpful.
And sorry for the noise that I made.

**Comment by TrayanZapryanov** [comment]
@mgravell What do you think about following extensions?

```
internal static class DictionaryExtensions
{
	public static bool TryGet<TValue>(this Dictionary<string, TValue> dictionary, KeyBuilder key, out TValue value)
	{
		if (dictionary.TryGetAlternateLookup(out Dictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.TryGetValue(key.GetSpan(), out value);
			key.Reset();
			return result;
		}

		return dictionary.TryGetValue(key.GetSpan().ToString(), out value);
	}

	public static bool TryAdd<TValue>(this Dictionary<string, TValue> dictionary, KeyBuilder key, TValue value)
	{
		if (dictionary.TryGetAlternateLookup(out Dictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.TryAdd(key.GetSpan(), value);
			key.Reset();
			return result;
		}

		return dictionary.TryAdd(key.GetSpan().ToString(), value);
	}

	public static bool ContainsKey<TValue>(this Dictionary<string, TValue> dictionary, KeyBuilder key)
	{
		if (dictionary.TryGetAlternateLookup(out Dictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.ContainsKey(key.GetSpan());
			key.Reset();
			return result;
		}

		return dictionary.ContainsKey(key.GetSpan().ToString());
	}

	public static bool Remove<TValue>(this Dictionary<string, TValue> dictionary, KeyBuilder key)
	{
		if (dictionary.TryGetAlternateLookup(out Dictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.Remove(key.GetSpan());
			key.Reset();
			return result;
		}

		return dictionary.Remove(key.GetSpan().ToString());
	}
}
```
Same for ConcurrentDictionary

```
internal static class ConcurrentDictionaryExtensions
{
	public static bool TryGetValue<TValue>(this ConcurrentDictionary<string, TValue> dictionary, KeyBuilder key, out TValue value)
	{
		if (dictionary.TryGetAlternateLookup(out ConcurrentDictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.TryGetValue(key.GetSpan(), out value);
			key.Reset();
			return result;
		}

		return dictionary.TryGetValue(key.GetSpan().ToString(), out value);
	}

	public static bool TryAdd<TValue>(this ConcurrentDictionary<string, TValue> dictionary, KeyBuilder key, TValue value)
	{
		if (dictionary.TryGetAlternateLookup(out ConcurrentDictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.TryAdd(key.GetSpan(), value);
			key.Reset();
			return result;
		}

		return dictionary.TryAdd(key.GetSpan().ToString(), value);
	}

	public static bool ContainsKey<TValue>(this ConcurrentDictionary<string, TValue> dictionary, KeyBuilder key)
	{
		if (dictionary.TryGetAlternateLookup(out ConcurrentDictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.ContainsKey(key.GetSpan());
			key.Reset();
			return result;
		}

		return dictionary.ContainsKey(key.GetSpan().ToString());
	}

	public static bool TryRemove<TValue>(this ConcurrentDictionary<string, TValue> dictionary, KeyBuilder key, out TValue value)
	{
		if (dictionary.TryGetAlternateLookup(out ConcurrentDictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> span))
		{
			var result = span.TryRemove(key.GetSpan(), out value);
			key.Reset();
			return result;
		}

		return dictionary.TryRemove(key.GetSpan().ToString(), out value);
	}
}
```

They seem quite useful and if they are added to the .NET, once compiled will provide allocation free operations.
I've added one more test 

```
private Guid TryGetBySbcSpanExtension(string region, int id)
{
	ConcurrentDictionaryExtensions.TryGetValue(strings, $"/foos/{region}/{id}", out var value);
	return value;
}

[Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
public void SBC_Span_TryGetExtension()
{
	foreach (var value in testValues)
	{
		_ = TryGetBySbcSpanExtension(value.region, value.id);
	}
}
```
 And here it is results from my machine:
BenchmarkDotNet v0.14.0, Windows 11 (10.0.26100.3037)
11th Gen Intel Core i9-11900K 3.50GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 9.0.103
  [Host]     : .NET 9.0.2 (9.0.225.6610), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  DefaultJob : .NET 9.0.2 (9.0.225.6610), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


| Method                             | Mean     | Error    | StdDev   | Gen0   | Allocated |
|----------------------------------- |---------:|---------:|---------:|-------:|----------:|
| Tuple                              | 14.11 ns | 0.212 ns | 0.188 ns |      - |         - |
| String                             | 40.64 ns | 0.554 ns | 0.519 ns | 0.0067 |      56 B |
| DISH_Span                          | 40.65 ns | 0.790 ns | 0.739 ns |      - |         - |
| StackAlloc_DISH_Span               | 36.00 ns | 0.169 ns | 0.141 ns |      - |         - |
| StackAllocSkipLocalsInit_DISH_Span | 35.72 ns | 0.150 ns | 0.125 ns |      - |         - |
| SBC_Span                           | 38.63 ns | 0.724 ns | 0.804 ns |      - |         - |
| SBC_Span_TryGetExtension           | 31.30 ns | 0.620 ns | 0.580 ns |      - |         - |
